vars/selfUnitTest.groovy

```diff
@@ -0,0 +1,112 @@
+/* groovylint-disable ParameterName */
+// vars/selfUnitTest.groovy
+
+/**
+ * selfUnitTest.groovy
+ *
+ * Runs self unit tests.
+ */
+
+
+void call() {
+    // Save and restore env.pragmas after testing
+    env_pragmas_original = env.pragmas
+    try {
+        _test_pragmas_env()
+    } finally {
+        env.pragmas = env_pragmas_original
+    }
+}
+
+void _test_pragmas_env() {
+    commit_message = '''Skip-build: true
+Skip-PR-comments: true
+
+My-pragma1:   val1
+My-pragma2:  val2 val2  
+
+Required-githooks: true
+
+Signed-off-by: Brian J. Murrell <brian.murrell@intel.com>'''
+
+    Map expected_map = [
+        "skip-build": "true", "skip-pr-comments": "true",
+        "my-pragma1": "val1", "my-pragma2": "val2 val2",
+        "required-githooks": "true",
+        "signed-off-by": "Brian J. Murrell <brian.murrell@intel.com>"]
+    String expected_str = '{skip-build=true, skip-pr-comments=true, ' +
+                          'my-pragma1=val1, my-pragma2=val2 val2, ' +
+                          'required-githooks=true, signed-off-by=Brian J. Murrell <brian.murrell@intel.com>}'
+
+    println("Test pragmasToMap")
+
+    println("  with commit message")
+    result_map = pragmasToMap(commit_message)
+    println("    result_map   = ${result_map}")
+    println("    expected_map = ${expected_map}")
+    assert(result_map == expected_map)
+
+    println("  with empty string")
+    result_map = pragmasToMap("")
```

Comment:
Maybe ditto here?  So perhaps we need the commit message in parts.  One static part for title and description and we add the pragmas, or no pragmas or garbage to that in each test?

Reply:
I don't follow here. Do you mean have multiple arguments to the pragmasToMap function?

Reply:
Or do you mean test cases for each of these, for example?
```
DAOS-XXXX test: my commit title
```

```
DAOS-XXXX test: my commit title

<pragmas>
```

```
some garbage
```

Reply:
> I don't follow here. Do you mean have multiple arguments to the pragmasToMap function?

I just mean that given that pragamsToMap() takes a commit message as it's argument, having it empty here doesn't represent what it will see when being used.

I was just further suggesting that for ease, you might want to build the commit messages passed to these various tests from a static _Tltle/Description_ string you define once and then append another string with the various pragmas you want to test.  Just to reduce the duplication of the Title/Description part.

Reply:
> having it empty here doesn't represent what it will see when being used.

In practice, yes. But testing with an empty commit message validates the code path where `pragamsToMap` is given an empty string. Part of that test stems from me seeing different behavior with this Jenkins/groovy vs my local groovy install.

> build the commit messages passed ...

Sorry, yes, I understand what you mean now :)

Reply:
> In practice, yes. But testing with an empty commit message validates the code path where `pragamsToMap` is given an empty string.

But given that `pragmasToMap()` is given a _commit message_, when could that ever be empty?  I don't think you can create a git commit without any message at all.

Reply:
In practice, it won't. But the function implementation itself still supports that code path. So having a test for it verifies the function behaves as expected. IMO, it doesn't matter if it's used in practice or not. If the code path is possible, it should ideally be tested. And the fewer assumptions this function has to make about a commit message, the more robust and futureproof it is

Reply:
> In practice, it won't. But the function implementation itself still supports that code path.

That's a good point.  But then we should ask should it?  Given that we can assume that a commit message should never be empty, if that function gets an empty commit message argument, that ought to be an error by the caller, and called out.  Maybe that function should `error()` if the argument is empty.  We probably don't do enough of that kind of argument checking in the library, but we should.

Reply:
I've been erring on the side of caution here and trying to minimally change behavior. If we want this to be an error, can we do that separately? TBH I'm not familiar with groovy `error()` so it would probably take me a few cycles to figure out the right code+test

Reply:
> I've been erring on the side of caution here and trying to minimally change behavior.

I don't think that adding validation that should have been there in the first place is necessarily changing behavior.  It's only becoming obvious it's missing as you are adding the testing now and adding test cases for what shouldn't even be considered valid input.  So probably the test case is still valid but just tests that the function does `error()` out.

> If we want this to be an error, can we do that separately? TBH I'm not familiar with groovy `error()` so it would probably take me a few cycles to figure out the right code+test

`error()` is not groovy per-se but is a Jenkins step function.  It's simply `error('This function requires a valid commit message as it's first argument')`.  That's it.  That stops the pipeloine at that point, displays the error and bails.

So for the purposes of testing, you would need to mock it.  I believe you will find an example of that in some of the (probably commented out) U/T hackery that has been added to (some) other function(s).

Reply:
> for the purposes of testing, you would need to mock it

That's exactly the kind of rabbit hole I'm trying to avoid here. I've already spent _**a lot more**_ time on this than I thought I needed, and have already fixed multiple existing issues.

Reply:
> That's exactly the kind of rabbit hole I'm trying to avoid here. I've already spent _**a lot more**_ time on this than I thought I needed, and have already fixed multiple existing issues.

I can certainly understand that.  But TBH, the argument validation (for a completely empty commit) and the resulting `error()` if it is empty seems straightforward enough to not really even need a test case, IMO.

Reply:
If we're okay with not having a test case, then sure. Just updated